File: tests/basic/function_beta.ml
open Optitrust
open Target


let _ = Run.doc_script_cpp (fun _ ->

     Variable_basic.inline ~accept_functions:true [cFunDef "sq"];
  !! Function_basic.beta [cVarDef "r"; cFun ""];

)

"
int sq(int x) { return (x * x); }

int main() {
  int r = sq(3);
}
"

let _ = Run.script_cpp (fun _ ->
  (* Functions *)
  !! Variable_basic.unfold ~accept_functions:true [cTopFunDef "f"];
  !! Function_basic.beta [cTopFunDef "test_fun"; cFun ""];

  !! Debug_transfo.current_ast_at_target ~style:InternalDisplay "f_X def" [cFunDef "f_X"];
  !! Debug_transfo.current_ast_at_target ~style:InternalDisplay "f_X call" [cTopFunDef "test_method"; cFun ""];

  (* Class methods *)
  !! Variable_basic.unfold ~accept_functions:true [cFunDef "f_X"];
  !! Function_basic.beta [cTopFunDef "test_method"; cFun ""];

  (* Namespaces *)
  (* FIXME:
    !! Variable_basic.unfold ~accept_functions:true [cFunDef "h"];
    !! Function_basic.beta [cTopFunDef "g"; cFun ""]; *)
)

File: tests/basic/function_beta.cpp
void f(int j) {
  int s = 0;
  s += 2*j;
  s -= j;

}

void test_fun(){
  int i = 1;
  f(i);

}

class X {

  int x;
  public:
    int f_X(int y){
      return x + y;
    }
};

void test_method(){
  int a = 1;
  X b;
  int c = b.f_X(a);

}

namespace Y {
  void h(int& a) {
    a = 1;
  }
}

namespace Z {
  void g(int a, int, int) {
    Y::h(a);
  }
}

int main () {
  int x = 1;
  Y::h(x);
}
